lib/cn/kcompact.c

```diff
@@ -175,176 +79,209 @@ kcompact(struct cn_compaction_work *w, struct kvset_builder *bldr)
     if (w->cw_prog_interval && w->cw_progress)
         tprog = jiffies;
 
-    err = merge_init(&bh, w->cw_inputv, w->cw_kvset_cnt, &w->cw_stats);
+    err = bin_heap2_create(w->cw_kvset_cnt, kv_item_compare, &bh);
     if (ev(err))
         return err;
 
-    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
-    if (!more || ev(err))
+    sources = malloc(w->cw_kvset_cnt * sizeof(*sources));
+    if (!sources) {
+        err = merr(ENOMEM);
         goto done;
+    }
 
-new_key:
+    for (uint i = 0; i < w->cw_kvset_cnt; i++) {
+        struct kv_iterator *iter = w->cw_inputv[i];
 
-    if (atomic_read(w->cw_cancel_request)) {
-        err = merr(ev(ESHUTDOWN));
-        goto done;
+        sources[i] = kvset_iter_es_get(iter);
     }
 
-    if (tprog) {
-        u64 now = jiffies;
+    err = bin_heap2_prepare(bh, w->cw_kvset_cnt, sources);
+    if (ev(err))
+        goto done;
+    /* In the event this assert fails, at least one iterator is EOF and the idea
+     * that struct element_source::es_sort will properly index the vblock map is
+     * incorrect. Kvsets won't typically exist if they have no prefix tombstones
+     * or keys.
+     */
+    assert(bin_heap2_width(bh) == w->cw_kvset_cnt);
```

Comment:
Might be safer to also add a live check in here and return EBUG? If we ever proceed further when this condition is false on a build with asserts disabled it can lead to DB corruption, I think...

Reply:
Hmmm. This would be a huge logic error if we had a kvset that couldn't produce at least a single key. Ideally it would have been caught earlier up the stack.

Reply:
Fair point. If we have checks for that earlier up the stack then adding one here would be redundant..